routers/web/repo/release.go

```diff
@@ -118,9 +118,12 @@ func releasesOrTags(ctx *context.Context, isTagList bool) {
 
 	opts := repo_model.FindReleasesOptions{
 		ListOptions:   listOptions,
-		IncludeDrafts: writeAccess && !isTagList,
+		IncludeDrafts: isTagList || writeAccess,
```

Comment:
Is it correct? If I understand correctly, the `writeAccess && ...` means that the old logic didn't want to show drafts to users who do not have write access.

Even if in the end the result is right, it seems pretty hard to understand why it works well.

Reply:
Or maybe separate the logic for `isTagList`, and use enum for IncludeTag (No/HasCommitID/All)

```
if isTagList {
    IncludeDrafts: true
    IncludeTag: HasCommitID
} else {
    IncludeDrafts: writeAccess
    IncludeTag: No
}
```

```
FindReleasesOptions.toCond() {
    if IncludeTag == No {
    } else if IncludeTag == HasCommitID {
    } else if IncludeTag == All {
    }
}
```

Reply:
The point is that the `writeAccess` should only work with listing releases, it has nothing to do with `writeAccess` if it is to list tags. The old logic is correct only if there are no draft releases with existing tags.

The new logic is that: if the `isTagList` is true, the `IncludeDrafts` should be true, and with a restriction: only drafts with existing tags. So I think it is clear enough now.

I understand your idea, but I believe what you mean is using enum for `IncludeDrafts`, not `IncludeTags`. Like:

```go
if isTagList {
    IncludeDrafts: HasCommitID
    IncludeTags: true
} else {
   IncludeDrafts: All (if writeAccess) or No
   IncludeTags: false
}
```

It's definitely going to work, but it may not be worth defining a new enum type, what do you think?

BTW,  I do think `HasCommitID` is a better name, but it should not be used here, because the `release` model doesn't have a field named `CommitID`, only `Sha1`. Perhaps it is the problem of the `release` model, in fact, I do think there are some tricky designs in it, which make it difficult to list tags correctly. Anyway, refactoring is out of scope for this PR.

Reply:
Without introducing more enum, the `if` could be like this with more comments:


```go
	opts := repo_model.FindReleasesOptions{
		ListOptions:   listOptions,
	}
	if isTagList {
		// for the Tags list page, show all releases with real tags (having real commit-id), the drafts should also be included because a real tag might be used as a draft
		opts.IncludeDrafts = true
		opts.IncludeTags = true
		opts.HasSha1 = true
	} else {
		// only show draft releases for users who can write, read-only users shouldn't see draft releases.
		opts.IncludeDrafts = writeAccess
	}
```

Reply:
Looks better, I'll take it.